library/std/src/sys/sgx/entropy.rs

```diff
@@ -0,0 +1,36 @@
+use crate::io::{const_io_error, default_read, BorrowedCursor, ErrorKind, Read, Result};
+
+pub const INSECURE_HASHMAP: bool = false;
+
+pub struct Entropy {
+    pub insecure: bool,
+}
+
+impl Read for Entropy {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        default_read(self, buf)
+    }
+
+    fn read_buf(&mut self, mut buf: BorrowedCursor<'_>) -> Result<()> {
+        if buf.capacity() != 0 {
+            let rand = rdrand64()
+                .ok_or(const_io_error!(ErrorKind::WouldBlock, "no random data available"))?;
+            buf.append(&rand.to_ne_bytes()[..usize::min(buf.capacity(), 8)]);
+            Ok(())
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub(super) fn rdrand64() -> Option<u64> {
+    unsafe {
+        let mut ret: u64 = 0;
+        for _ in 0..10 {
+            if crate::arch::x86_64::_rdrand64_step(&mut ret) == 1 {
```

Comment:
Not a maintainer of sgx but this does not offer high quality entropy. See the "Generating seeds with rdrand" section in https://www.intel.com/content/www/us/en/developer/articles/guide/intel-digital-random-number-generator-drng-software-implementation-guide.html for how to do it (which we probably don't want in std...)

Reply:
I... don't quite get that section, unfortunately.

IIUC (please feel free to correct me), `RDRAND` is like `/dev/urandom` in that it uses a CSPRNG to generate more random data from a single, truly random seed. Because of the size of the seed, constant reseeding and the strength of the PRNG, the output should not be in any practical way less secure that `RDSEED`. But that doesn't explain why anyone would use `RDSEED`...

I didn't change this, because [`getrandom` uses `RDRAND` on SGX](https://github.com/rust-random/getrandom/blob/2f9989fba13f4b1e836ef1d1d614f047d0b110e6/src/lib.rs#L268-L269). Note that it does not consider values `0` or `u64::MAX` valid because of hardware bugs on AMD devices. Since SGX is Intel-only, I guess it isn't a concern here?

If `RDRAND` is truly the wrong option here, I guess the alternative is doing what I did for Hermit [here](https://github.com/hermitcore/libhermit-rs/pull/640) and use `RDSEED` to seed our own PRNG.

CC @raoulstrackx

Reply:
> ...the DRNG using the RDRAND instruction is useful for generating high-quality keys for cryptographic protocols, and the RSEED instruction is provided for seeding software-based pseudorandom number generators (PRNGs)

I think that sums it up nicely. If someone needs stronger guarantees for a PRNG, they should take a look at what exactly happens in `getrandom`. The documentation explicitly mentions the randomness comes from `rdrand`. To me that looks fine, but I'm not a cryptographic. @zugzwang can you pitch in?

Indeed, AMD hardware bugs have no impact on SGX and the values `0` and `u64::MAX` should not be handled differently.

Reply:
Makes sense to me. Intel documentation looks clear, and there is useful discussion on security levels [here](https://crypto.stackexchange.com/questions/42340/usage-difference-between-x86-rdrand-and-rdseed).